### PR TITLE
Add Query Collector Module

### PR DESCRIPTION
**Pull request by SuhasHebbar**
The Query Collector is designed such that on receiving a set (set as max_itineraries) number of itineraries, it will send the completed itineraries obtained to the query collector and terminate.

In case the number of valid itineraries for the query is less less than max_itineraries then a timeout has been added such that if the Query Collector does not receive any message for a period of time (set as timeout) then it will send the Itineraries obtained till that point to UQC and terminate.

### CODE REVIEW

File: lib/ts/station/query_collector.ex [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    itinerary_fn = dependency.itinerary
+
+    qid = itinerary_fn.get_query_id(itinerary)
+    Logger.debug(fn -> "Unregistering query: #{qid}" end)
+
+    # Unregister query using registry
+    registry.unregister_query(qid)
+  end
+
+  def handle_cast(:search, {itinerary, opts, dependency}) do
```
Comment (prasadtalasila) [inline lib/ts/station/query_collector.ex]: Appropriate atom could be `:start` or `:begin`

File: lib/ts/station/registry.ex [hunk 2]
```diff
@@ -40,6 +40,24 @@ defmodule Station.Registry do
     end
   end
 
+    # Returns the Query Collecotr pid associated with the query id
+    def lookup_query_id(qid) do
```
Comment (prasadtalasila) [inline lib/ts/station/registry.ex]: unit tests are needed for this function.

**Review by prasadtalasila — CHANGES_REQUESTED** [review]

File: lib/ts/station/registry.ex (hunk 2 above)
Comment (SuhasHebbar) [reply lib/ts/station/registry.ex]: Test "Perform lookup for registered and unregistered query" has been added.

File: test/ts/collector_test.exs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+    itinerary = Itinerary.new(query, preference)
+
+    # Set options for QC process
+    opts = %{max_itineraries: 5, timeout: 100}
+
+    dependency = @dependency
+
+    # Get 5 random terminal itineraries of length 4
+    itinerary1 = get_terminal_itinerary(query, preference, 4)
```
Comment (prasadtalasila) [inline test/ts/collector_test.exs:40]: reformat query to contain query_id and itinerary.

Comment (prasadtalasila) [reply test/ts/collector_test.exs:40]: Rename the existing itinerary variable to `itinerary_acc = {query, itinerary, preference}`. Please make this change in all the modules.

File: test/ts/collector_test.exs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    itinerary2 = get_terminal_itinerary(query, preference, 4)
+    itinerary3 = get_terminal_itinerary(query, preference, 4)
+    itinerary4 = get_terminal_itinerary(query, preference, 4)
+    itinerary5 = get_terminal_itinerary(query, preference, 4)
+
+    result = [itinerary5, itinerary4, itinerary3, itinerary2, itinerary1]
```
Comment (prasadtalasila) [inline test/ts/collector_test.exs]: QC would remove redundant information.

**Review by prasadtalasila — CHANGES_REQUESTED** [review]
@SuhasHebbar Please make the suggested changes.

**Comment by prasadtalasila** [comment]
We also need logger statements to help calculate the query processing time.

**Review by prasadtalasila — APPROVED** [review]

**Pull request merged**